quic/s2n-quic-core/src/buffer/reassembler.rs

```diff
@@ -158,115 +166,230 @@ impl Reassembler {
+    fn check_reader_fin<R>(&mut self, reader: &mut R) -> Result<(), Error<R::Error>>
+    where
+        R: Reader + ?Sized,
+    {
+        let buffered_offset = reader
+            .current_offset()
+            .checked_add_usize(reader.buffered_len())
+            .ok_or(Error::OutOfRange)?
+            .as_u64();
 
-        // make sure the request isn't trying to write beyond the final size
         //= https://www.rfc-editor.org/rfc/rfc9000#section-4.5
         //# Once a final size for a stream is known, it cannot change.  If a
         //# RESET_STREAM or STREAM frame is received indicating a change in the
         //# final size for the stream, an endpoint SHOULD respond with an error
         //# of type FINAL_SIZE_ERROR; see Section 11 for details on error
         //# handling.
-        ensure!(excess.is_empty(), Err(Error::InvalidFin));
+        match (reader.final_offset(), self.final_size()) {
+            (Some(actual), Some(expected)) => {
+                ensure!(actual == expected, Err(Error::InvalidFin));
+            }
+            (Some(final_offset), None) => {
+                let final_offset = final_offset.as_u64();
 
-        // if the request is empty we're done
-        ensure!(!request.is_empty(), Ok(()));
+                // make sure that we didn't see any previous chunks greater than the final size
+                ensure!(
+                    self.cursors.max_recv_offset <= final_offset,
+                    Err(Error::InvalidFin)
+                );
+
+                self.cursors.final_offset = final_offset;
+            }
+            (None, Some(expected)) => {
+                // make sure the reader doesn't exceed a previously known final offset
+                ensure!(expected >= buffered_offset, Err(Error::InvalidFin));
+            }
+            (None, None) => {}
+        }
 
         // record the maximum offset that we've seen
-        self.max_recv_offset = self.max_recv_offset.max(request.end_exclusive());
+        self.cursors.max_recv_offset = self.cursors.max_recv_offset.max(buffered_offset);
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn write_reader_impl<R>(&mut self, reader: &mut R) -> Result<(), R::Error>
+    where
+        R: Reader + ?Sized,
+    {
+        // if the reader is empty at this point, just make sure it doesn't return an error
+        if reader.buffer_is_empty() {
+            let _chunk = reader.read_chunk(0)?;
+            return Ok(());
+        }
+
+        let mut selected = None;
 
         // start from the back with the assumption that most data arrives in order
-        for mut idx in (0..self.slots.len()).rev() {
-            unsafe {
-                assume!(self.slots.len() > idx);
-            }
-            let slot = &mut self.slots[idx];
+        for idx in (0..self.slots.len()).rev() {
+            let Some(slot) = self.slots.get(idx) else {
+                debug_assert!(false);
+                unsafe { core::hint::unreachable_unchecked() }
```

Comment:
what makes this unreachable?

Reply:
we're using the `idx` above, which is created with `0..slots.len()`, so `idx` will always be in bounds. I initially had an iterator using the `VecDeque::iter` and, oddly enough, it ended up being about 5% slower than the regular integers.

I can add a safety comment for clarity.